crates/bitwarden-threading/src/thread_bound_runner.rs

```diff
@@ -0,0 +1,228 @@
+#![allow(dead_code)]
+#![allow(unused_variables)]
+
+use std::{future::Future, pin::Pin, rc::Rc};
+
+use bitwarden_error::bitwarden_error;
+use thiserror::Error;
+#[cfg(not(target_arch = "wasm32"))]
+use tokio::task::spawn_local;
+#[cfg(target_arch = "wasm32")]
+use wasm_bindgen_futures::spawn_local;
+
+type CallFunction<ThreadState> =
+    Box<dyn FnOnce(Rc<ThreadState>) -> Pin<Box<dyn Future<Output = ()>>> + Send>;
+
+struct CallRequest<ThreadState> {
+    function: CallFunction<ThreadState>,
+}
+
+/// The call failed before it could return a value. This should not happen unless
+/// the thread panics, which can only happen if the function passed to `run_in_thread`
+/// panics.
+#[derive(Debug, Error)]
+#[error("The call failed before it could return a value: {0}")]
+#[bitwarden_error(basic)]
+pub struct CallError(String);
+
+/// A runner that takes a non-`Send`, non-`Sync` state and makes it `Send + Sync` compatible.
+///
+/// `ThreadBoundRunner` is designed to safely encapsulate a `!Send + !Sync` state object by
+/// pinning it to a single thread using `spawn_local`. It provides a `Send + Sync` API that
+/// allows other threads to submit tasks (function pointers or closures) that operate on the
+/// thread-bound state.
+///
+/// Tasks are queued via an internal channel and are executed sequentially on the owning thread.
+///
+/// # Example
+/// ```ignore
+/// let runner = ThreadBoundRunner::new(my_state);
+///
+/// runner.run_in_thread(|state| async move {
+///     // do something with `state`
+/// });
+/// ```
+///
+/// This pattern is useful for interacting with APIs or data structures that must remain
+/// on the same thread, such as GUI toolkits, WebAssembly contexts, or other thread-bound
+/// environments.
+#[derive(Clone)]
+pub struct ThreadBoundRunner<ThreadState> {
+    call_channel_tx: tokio::sync::mpsc::Sender<CallRequest<ThreadState>>,
+}
+
+impl<ThreadState> ThreadBoundRunner<ThreadState>
+where
+    ThreadState: 'static,
+{
+    pub fn new(state: ThreadState) -> Self {
+        let (call_channel_tx, mut call_channel_rx) =
+            tokio::sync::mpsc::channel::<CallRequest<ThreadState>>(1);
+
+        spawn_local(async move {
+            let state = Rc::new(state);
+            while let Some(request) = call_channel_rx.recv().await {
+                let state = state.clone();
+                spawn_local(async move {
+                    (request.function)(state).await;
+                });
```

Comment:
This can probably be simplified a little, as the function is already a future, but it's not a big deal either way:
```suggestion
                spawn_local((request.function)(state.clone()));
```